django/dispatch/dispatcher.py

```diff
@@ -108,12 +110,23 @@ def connect(self, receiver, sender=None, weak=True, dispatch_uid=None):
                 ref = weakref.WeakMethod
                 receiver_object = receiver.__self__
             receiver = ref(receiver)
-            weakref.finalize(receiver_object, self._remove_receiver)
+            weakref.finalize(receiver_object, self._flag_dead_receivers)
+
+        # Keep a weakref to sender if possible to ensure associated
+        # receivers are cleared if it gets garbage collected. This ensures
+        # there is no id(sender) collisions for distinct senders with
+        # non-overlapping lifetimes.
+        sender_ref = None
+        if sender is not None:
+            try:
+                sender_ref = weakref.ref(sender, self._flag_dead_receivers)
+            except TypeError:
+                pass
```

Comment:
We could emit a runtime or start a deprecation cycle if we wanted to let user know about the risk of providing non-weakrefable senders that can potentially be garbage collected. This seemed unlikely enough that no changes were warranted in my opinion.